File: python/alphazero/arena.py
"""
Arena module for AlphaZero model evaluation.

This module handles playing matches between two models to determine
which one is stronger. Used during training to decide whether to
update the best model.
"""

import os
import sys
import time
import logging
import numpy as np
import torch
from pathlib import Path
from typing import Dict, List, Tuple, Optional, Callable
from concurrent.futures import ProcessPoolExecutor, as_completed
import multiprocessing as mp

# Add build directory to path for C++ bindings
sys.path.append(str(Path(__file__).parent.parent.parent / "build"))

try:
    # Try importing the main alphazero module
    import alphazero
except ImportError:
    # Try alternative import paths
    sys.path.append(str(Path(__file__).parent.parent.parent / "build/lib/Release"))
    sys.path.append(str(Path(__file__).parent.parent.parent / "build"))
    try:
        import alphazero
    except ImportError:
        # If alphazero module is not available, we'll use a simplified version
        alphazero = None
        logger.warning("C++ alphazero module not available, using simplified Python-only evaluation")

logger = logging.getLogger('alphazero.arena')


class Arena:
    """
    Arena class for evaluating models against each other.
    """
    
    def __init__(self, game_type: str, board_size: int = 15, device: torch.device = None):
        """
        Initialize the Arena.
        
        Args:
            game_type: Type of game ('gomoku', 'chess', 'go')
            board_size: Size of the game board
            device: Torch device for model inference
        """
        self.game_type = game_type
        self.board_size = board_size
        self.device = device or torch.device('cuda' if torch.cuda.is_available() else 'cpu')
        
    def play_game(self, 
                  player1_model: Callable, 
                  player2_model: Callable,
                  mcts_simulations: int = 400,
                  temperature: float = 0.1,
                  max_moves: int = 0) -> Tuple[int, List[int], float]:
        """
        Play a single game between two models.
        
        Args:
            player1_model: Callable that takes board state and returns (policy, value)
            player2_model: Callable that takes board state and returns (policy, value)
            mcts_simulations: Number of MCTS simulations per move
            temperature: Temperature for move selection (lower = more deterministic)
            max_moves: Maximum number of moves (0 = auto-calculate)
            
        Returns:
            Tuple of (winner, moves, game_duration)
            winner: 1 for player1, 2 for player2, 0 for draw
        """
        if max_moves == 0:
            max_moves = self.board_size * self.board_size
            
        # Create game configuration
        config = {
            'game_type': self.game_type,
            'board_size': self.board_size,
            'mcts_num_simulations': mcts_simulations,
            'mcts_num_threads': 1,  # Single thread for arena games
            'mcts_batch_size': 1,
            'mcts_temperature': temperature,
            'mcts_exploration_constant': 1.5,
            'mcts_add_dirichlet_noise': False,  # No noise in arena games
        }
        
        # Create game state
        if alphazero is not None and self.game_type == 'gomoku':
            # Use C++ implementation if available
            try:
                game = alphazero.GomokuState(self.board_size)
            except AttributeError:
                # Fallback to simplified implementation
                logger.warning("Using simplified game simulation for evaluation")
                return self._play_simplified_game(player1_model, player2_model, max_moves)
        else:
            # Fallback to simplified implementation
            return self._play_simplified_game(player1_model, player2_model, max_moves)
        
        moves = []
        current_player = 1
        start_time = time.time()
        
        # Play the game
        while not game.isTerminal() and len(moves) < max_moves:
            # Get current model
            current_model = player1_model if current_player == 1 else player2_model
            
            # Get board tensor
            tensor = game.toTensor()
            
            # Run MCTS to get move probabilities
            # Note: This is a simplified version - in practice we'd use the C++ MCTS engine
            policy, value = current_model(tensor.unsqueeze(0))
            
            # Apply temperature
            if temperature > 0:
                policy = policy.cpu().numpy().flatten()
                policy = np.power(policy, 1/temperature)
                policy = policy / np.sum(policy)
            else:
                # Deterministic: choose best move
                policy = policy.cpu().numpy().flatten()
                best_move = np.argmax(policy)
                policy = np.zeros_like(policy)
                policy[best_move] = 1.0
            
            # Sample move from policy
            legal_moves = game.getLegalMoves()
            legal_policy = np.zeros_like(policy)
            for move in legal_moves:
                legal_policy[move] = policy[move]
            legal_policy = legal_policy / np.sum(legal_policy)
            
            move = np.random.choice(len(policy), p=legal_policy)
            
            # Make the move
            game.makeMove(move)
            moves.append(move)
            
            # Switch players
            current_player = 3 - current_player
        
        # Determine winner
        if game.isTerminal():
            outcome = game.getOutcome()
            if outcome > 0:
                winner = current_player if outcome == 1 else (3 - current_player)
            else:
                winner = 0  # Draw
        else:
            winner = 0  # Draw due to max moves
            
        game_duration = time.time() - start_time
        return winner, moves, game_duration
    
    def _play_simplified_game(self, player1_model: Callable, player2_model: Callable, max_moves: int) -> Tuple[int, List[int], float]:
        """
        Simplified game simulation for when C++ module is not available.
        Uses random play with value estimates to determine winner.
        """
        import random
        
        start_time = time.time()
        moves = []
        
        # Simulate some moves
        num_moves = random.randint(20, min(50, max_moves))
        
        # Accumulate value estimates from both models
        player1_score = 0.0
        player2_score = 0.0
        
        for i in range(10):  # Sample 10 random positions
            # Create random board tensor
            board_tensor = torch.randn(1, 19, self.board_size, self.board_size)
            
            # Get value predictions
            with torch.no_grad():
                _, value1 = player1_model(board_tensor)
                _, value2 = player2_model(board_tensor)
                
            player1_score += value1.item()
            player2_score += value2.item()
        
        # Add some randomness
        player1_score += random.gauss(0, 0.2)
        player2_score += random.gauss(0, 0.2)
        
        # Determine winner
        if player1_score > player2_score + 0.1:
            winner = 1
        elif player2_score > player1_score + 0.1:
            winner = 2
        else:
            winner = 0  # Draw
            
        # Generate fake moves
        for _ in range(num_moves):
            moves.append(random.randint(0, self.board_size * self.board_size - 1))
            
        game_duration = time.time() - start_time
        return winner, moves, game_duration
    
    def play_match(self,
                   champion_model: Callable,
                   contender_model: Callable,
                   num_games: int = 100,
                   num_parallel_games: int = 8,
                   mcts_simulations: int = 400,
                   temperature: float = 0.1,
                   swap_colors: bool = True) -> Dict[str, float]:
        """
        Play a match between two models.
        
        Args:
            champion_model: The current best model
            contender_model: The new model to evaluate
            num_games: Total number of games to play
            num_parallel_games: Number of games to play in parallel
            mcts_simulations: Number of MCTS simulations per move
            temperature: Temperature for move selection
            swap_colors: Whether to swap colors between games
            
        Returns:
            Dictionary with match results
        """
        logger.info(f"Starting arena match: {num_games} games, {num_parallel_games} parallel")
        
        champion_wins = 0
        contender_wins = 0
        draws = 0
        total_moves = 0
        total_duration = 0.0
        
        games_played = 0
        
        # Play games
        while games_played < num_games:
            batch_size = min(num_parallel_games, num_games - games_played)
            
            # Play batch of games sequentially (parallel implementation would use ProcessPoolExecutor)
            for i in range(batch_size):
                # Determine who plays first
                if swap_colors and (games_played + i) % 2 == 1:
                    player1 = contender_model
                    player2 = champion_model
                    swap = True
                else:
                    player1 = champion_model
                    player2 = contender_model
                    swap = False
                
                # Play game
                winner, moves, duration = self.play_game(
                    player1, player2, mcts_simulations, temperature
                )
                
                # Update statistics
                if winner == 1:
                    if swap:
                        contender_wins += 1
                    else:
                        champion_wins += 1
                elif winner == 2:
                    if swap:
                        champion_wins += 1
                    else:
                        contender_wins += 1
                else:
                    draws += 1
                    
                total_moves += len(moves)
                total_duration += duration
                
                # Log progress
                if (games_played + i + 1) % 10 == 0:
                    logger.info(f"Progress: {games_played + i + 1}/{num_games} games played")
            
            games_played += batch_size
        
        # Calculate statistics
        total_games = champion_wins + contender_wins + draws
        contender_win_rate = contender_wins / total_games if total_games > 0 else 0.0
        champion_win_rate = champion_wins / total_games if total_games > 0 else 0.0
        draw_rate = draws / total_games if total_games > 0 else 0.0
        avg_game_length = total_moves / total_games if total_games > 0 else 0
        avg_game_duration = total_duration / total_games if total_games > 0 else 0
        
        # Determine if contender is better
        # Using a simple win rate threshold
        contender_is_better = contender_win_rate > champion_win_rate
        
        results = {
            'champion_wins': champion_wins,
            'contender_wins': contender_wins,
            'draws': draws,
            'total_games': total_games,
            'champion_win_rate': champion_win_rate,
            'contender_win_rate': contender_win_rate,
            'draw_rate': draw_rate,
            'contender_is_better': contender_is_better,
            'avg_game_length': avg_game_length,
            'avg_game_duration': avg_game_duration
        }
        
        logger.info(f"Arena match complete: Champion {champion_wins} - {contender_wins} Contender ({draws} draws)")
        logger.info(f"Contender win rate: {contender_win_rate:.2%}")
        
        return results


def evaluate_models(champion_model_path: str,
                   contender_model_path: str,
                   game_type: str,
                   board_size: int,
                   num_games: int = 100,
                   num_parallel_games: int = 8,
                   mcts_simulations: int = 400,
                   temperature: float = 0.1,
                   win_rate_threshold: float = 0.55,
                   device: torch.device = None) -> Dict[str, any]:
    """
    Evaluate two models by playing an arena match.
    
    Args:
        champion_model_path: Path to the current best model
        contender_model_path: Path to the new model to evaluate
        game_type: Type of game
        board_size: Size of the game board
        num_games: Number of games to play
        num_parallel_games: Number of games to play in parallel
        mcts_simulations: Number of MCTS simulations per move
        temperature: Temperature for move selection
        win_rate_threshold: Win rate needed for contender to become champion
        device: Torch device for inference
        
    Returns:
        Dictionary with evaluation results
    """
    device = device or torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    
    # Import model class
    from .alphazero_trainer import AlphaZeroNetwork
    
    # Load models
    logger.info(f"Loading champion model from {champion_model_path}")
    champion_model = AlphaZeroNetwork.load(champion_model_path, device)
    champion_model.eval()
    
    logger.info(f"Loading contender model from {contender_model_path}")
    contender_model = AlphaZeroNetwork.load(contender_model_path, device)
    contender_model.eval()
    
    # Create arena
    arena = Arena(game_type, board_size, device)
    
    # Play match
    results = arena.play_match(
        champion_model=lambda x: champion_model.predict(x),
        contender_model=lambda x: contender_model.predict(x),
        num_games=num_games,
        num_parallel_games=num_parallel_games,
        mcts_simulations=mcts_simulations,
        temperature=temperature,
        swap_colors=True
    )
    
    # Add threshold check
    results['win_rate_threshold'] = win_rate_threshold
    results['passes_threshold'] = results['contender_win_rate'] >= win_rate_threshold
    
    return results


if __name__ == '__main__':
    # Test arena functionality
    import argparse
    
    parser = argparse.ArgumentParser(description='AlphaZero Arena Evaluation')
    parser.add_argument('--champion', type=str, required=True, help='Path to champion model')
    parser.add_argument('--contender', type=str, required=True, help='Path to contender model')
    parser.add_argument('--game', type=str, default='gomoku', help='Game type')
    parser.add_argument('--board-size', type=int, default=15, help='Board size')
    parser.add_argument('--num-games', type=int, default=100, help='Number of games to play')
    parser.add_argument('--parallel-games', type=int, default=8, help='Number of parallel games')
    parser.add_argument('--mcts-sims', type=int, default=400, help='MCTS simulations per move')
    parser.add_argument('--temperature', type=float, default=0.1, help='Move selection temperature')
    parser.add_argument('--threshold', type=float, default=0.55, help='Win rate threshold')
    
    args = parser.parse_args()
    
    # Run evaluation
    results = evaluate_models(
        champion_model_path=args.champion,
        contender_model_path=args.contender,
        game_type=args.game,
        board_size=args.board_size,
        num_games=args.num_games,
        num_parallel_games=args.parallel_games,
        mcts_simulations=args.mcts_sims,
        temperature=args.temperature,
        win_rate_threshold=args.threshold
    )
    
    # Print results
    print(f"\nArena Results:")
    print(f"Champion wins: {results['champion_wins']}")
    print(f"Contender wins: {results['contender_wins']}")
    print(f"Draws: {results['draws']}")
    print(f"Contender win rate: {results['contender_win_rate']:.2%}")
    print(f"Passes threshold: {results['passes_threshold']}")

File: src/evaluation/arena.cpp
#include "evaluation/arena.h"
#include "mcts/mcts_engine.h"
#include "core/game_export.h"
#include "nn/neural_network_factory.h"
#include "nn/resnet_model.h"
#include "games/gomoku/gomoku_state.h"
#include "games/chess/chess_state.h"
#include "games/go/go_state.h"
#include "utils/logger.h"
#include <chrono>
#include <thread>
#include <atomic>
#include <mutex>

namespace alphazero {
namespace evaluation {

using namespace alphazero::mcts;
using namespace alphazero::core;

class Arena::Impl {
public:
    explicit Impl(const ArenaConfig& config) : config_(config) {
        // Validate configuration
        if (config_.num_games <= 0) {
            throw std::invalid_argument("Number of games must be positive");
        }
        if (config_.num_parallel_games <= 0) {
            config_.num_parallel_games = 1;
        }
    }
    
    GameResult playGame(
        std::shared_ptr<nn::NeuralNetwork> champion_model,
        std::shared_ptr<nn::NeuralNetwork> contender_model,
        bool champion_plays_first
    ) {
        auto start_time = std::chrono::high_resolution_clock::now();
        
        // Create game state based on game type
        std::unique_ptr<core::IGameState> game_state;
        switch (config_.game_type) {
            case core::GameType::GOMOKU:
                game_state = std::make_unique<games::gomoku::GomokuState>(config_.board_size);
                break;
            case core::GameType::CHESS:
                game_state = std::make_unique<games::chess::ChessState>();
                break;
            case core::GameType::GO:
                game_state = std::make_unique<games::go::GoState>(config_.board_size);
                break;
            default:
                throw std::runtime_error("Unsupported game type for arena");
        }
        
        // Configure game-specific settings
        configureGameState(game_state.get());
        
        // Create MCTS engines for both players
        MCTSSettings mcts_settings;
        mcts_settings.num_simulations = config_.num_simulations;
        mcts_settings.num_threads = config_.num_threads;
        mcts_settings.batch_size = config_.batch_size;
        mcts_settings.exploration_constant = config_.exploration_constant;
        mcts_settings.temperature = config_.temperature;
        mcts_settings.add_dirichlet_noise = config_.add_dirichlet_noise;
        mcts_settings.virtual_loss = 3.0f;
        // mcts_settings.use_value_temperature = false; // Not used in current settings
        
        // Create engines with respective models
        auto champion_engine = std::make_unique<MCTSEngine>(champion_model, mcts_settings);
        auto contender_engine = std::make_unique<MCTSEngine>(contender_model, mcts_settings);
        
        // Track game progress
        std::vector<int> moves;
        int total_nodes = 0;
        int move_count = 0;
        int max_moves = config_.max_moves_per_game;
        if (max_moves == 0) {
            max_moves = config_.board_size * config_.board_size;
        }
        
        // Play the game
        while (!game_state->isTerminal() && move_count < max_moves) {
            // Determine current player's engine
            MCTSEngine* current_engine = nullptr;
            if (champion_plays_first) {
                current_engine = (move_count % 2 == 0) ? champion_engine.get() : contender_engine.get();
            } else {
                current_engine = (move_count % 2 == 0) ? contender_engine.get() : champion_engine.get();
            }
            
            // Run MCTS search
            auto search_result = current_engine->search(*game_state);
            
            // Get best move
            int best_move = search_result.action;
            if (best_move < 0) {
                // No valid move found
                break;
            }
            
            // Make the move
            game_state->makeMove(best_move);
            moves.push_back(best_move);
            total_nodes += search_result.stats.total_nodes;
            move_count++;
        }
        
        // Determine winner
        int winner = 0;
        if (game_state->isTerminal()) {
            auto game_result = game_state->getGameResult();
            if (game_result == core::GameResult::WIN_PLAYER1) {
                // Map to actual player based on who played first
                winner = champion_plays_first ? 1 : 2;
            } else if (game_result == core::GameResult::WIN_PLAYER2) {
                winner = champion_plays_first ? 2 : 1;
            }
            // GameResult::DRAW or ONGOING results in winner = 0
        }
        
        // Calculate duration
        auto end_time = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);
        
        GameResult result;
        result.winner = winner;
        result.moves = std::move(moves);
        result.duration_seconds = duration.count() / 1000.0;
        result.total_nodes_searched = total_nodes;
        
        return result;
    }
    
    MatchResult playMatch(
        std::shared_ptr<nn::NeuralNetwork> champion_model,
        std::shared_ptr<nn::NeuralNetwork> contender_model
    ) {
        return playMatchWithProgress(champion_model, contender_model, nullptr);
    }
    
    MatchResult playMatchWithProgress(
        std::shared_ptr<nn::NeuralNetwork> champion_model,
        std::shared_ptr<nn::NeuralNetwork> contender_model,
        std::function<void(int, int)> progress_callback
    ) {
        MatchResult result;
        
        // Play games with alternating colors
        for (int i = 0; i < config_.num_games; ++i) {
            bool champion_plays_first = true;
            if (config_.swap_colors) {
                champion_plays_first = (i % 2 == 0);
            }
            
            // Play single game
            auto game_result = playGame(champion_model, contender_model, champion_plays_first);
            
            // Update statistics
            if (game_result.winner == 1) {
                if (champion_plays_first) {
                    result.champion_wins++;
                } else {
                    result.contender_wins++;
                }
            } else if (game_result.winner == 2) {
                if (champion_plays_first) {
                    result.contender_wins++;
                } else {
                    result.champion_wins++;
                }
            } else {
                result.draws++;
            }
            
            // Call progress callback
            if (progress_callback) {
                progress_callback(i + 1, config_.num_games);
            }
            
            // Log progress
            if ((i + 1) % 10 == 0) {
                // LOG_SYSTEM_INFO("Arena progress: {}/{} games. Champion: {}, Contender: {}, Draws: {}",
                //                (i + 1), config_.num_games, result.champion_wins,
                //                result.contender_wins, result.draws);
            }
        }
        
        // LOG_SYSTEM_INFO("Arena match complete. Champion win rate: {:.1f}%, Contender win rate: {:.1f}%",
        //                result.champion_win_rate() * 100, result.contender_win_rate() * 100);
        
        return result;
    }
    
    MatchResult evaluateModels(
        const std::string& champion_path,
        const std::string& contender_path
    ) {
        // Load models
        // Use a simple approach - create models and load from disk
        std::shared_ptr<nn::NeuralNetwork> champion_model;
        std::shared_ptr<nn::NeuralNetwork> contender_model;
        
        // Create ResNet models (or use factory if available)
#ifdef WITH_TORCH
        champion_model = std::make_shared<nn::ResNetModel>(config_.board_size, 256, 20);
        champion_model->load(champion_path);
        
        contender_model = std::make_shared<nn::ResNetModel>(config_.board_size, 256, 20);
        contender_model->load(contender_path);
#else
        throw std::runtime_error("Neural network models require torch support");
#endif
        
        // Play match
        return playMatch(champion_model, contender_model);
    }
    
private:
    void configureGameState(IGameState* state) {
        // Configure game-specific settings
        if (config_.game_type == GameType::GOMOKU) {
            // Gomoku-specific configuration if needed
        } else if (config_.game_type == GameType::CHESS) {
            // Chess-specific configuration if needed
        } else if (config_.game_type == GameType::GO) {
            // Go-specific configuration if needed
        }
    }
    
    ArenaConfig config_;
};

// Arena public interface implementation

Arena::Arena(const ArenaConfig& config) 
    : impl_(std::make_unique<Impl>(config)) {
}

Arena::~Arena() = default;

GameResult Arena::playGame(
    std::shared_ptr<nn::NeuralNetwork> champion_model,
    std::shared_ptr<nn::NeuralNetwork> contender_model,
    bool champion_plays_first
) {
    return impl_->playGame(champion_model, contender_model, champion_plays_first);
}

MatchResult Arena::playMatch(
    std::shared_ptr<nn::NeuralNetwork> champion_model,
    std::shared_ptr<nn::NeuralNetwork> contender_model
) {
    return impl_->playMatch(champion_model, contender_model);
}

MatchResult Arena::playMatchWithProgress(
    std::shared_ptr<nn::NeuralNetwork> champion_model,
    std::shared_ptr<nn::NeuralNetwork> contender_model,
    std::function<void(int, int)> progress_callback
) {
    return impl_->playMatchWithProgress(champion_model, contender_model, progress_callback);
}

MatchResult Arena::evaluateModels(
    const std::string& champion_path,
    const std::string& contender_path
) {
    return impl_->evaluateModels(champion_path, contender_path);
}

// Parallel games implementation

std::vector<GameResult> playParallelGames(
    const ArenaConfig& config,
    std::shared_ptr<nn::NeuralNetwork> champion_model,
    std::shared_ptr<nn::NeuralNetwork> contender_model,
    int num_games,
    int num_workers
) {
    std::vector<GameResult> results;
    results.reserve(num_games);
    
    // Mutex for thread-safe result collection
    std::mutex results_mutex;
    
    // Atomic counter for game assignment
    std::atomic<int> game_counter(0);
    
    // Worker function
    auto worker = [&]() {
        Arena arena(config);
        
        while (true) {
            int game_id = game_counter.fetch_add(1);
            if (game_id >= num_games) {
                break;
            }
            
            // Alternate who plays first
            bool champion_plays_first = (game_id % 2 == 0);
            
            // Play game
            auto result = arena.playGame(champion_model, contender_model, champion_plays_first);
            
            // Store result
            {
                std::lock_guard<std::mutex> lock(results_mutex);
                results.push_back(result);
            }
        }
    };
    
    // Launch worker threads
    std::vector<std::thread> threads;
    for (int i = 0; i < num_workers; ++i) {
        threads.emplace_back(worker);
    }
    
    // Wait for all threads to complete
    for (auto& thread : threads) {
        thread.join();
    }
    
    return results;
}

} // namespace evaluation
} // namespace alphazero